numpy/core/einsumfunc.py

```diff
@@ -0,0 +1,718 @@
+        contraction = (contract_inds, False, einsum_str, input_list[:])
+        contraction_list.append(contraction)
+
+    opt_cost = sum(cost_list)
+
+    if einsum_call_arg:
+        return (operands, contraction_list)
+
+    # Return the path along with a nice string representation
+    overall_contraction = input_subscripts + "->" + output_subscript
+    header = ("scaling", "BLAS", "current", "remaining")
+
+    path_print  = "  Complete contraction:  %s\n" % overall_contraction
+    path_print += "         Naive scaling:  %2d\n" % len(indices)
+    path_print += "      Naive FLOP count:  %.3e\n" % naive_cost
+    path_print += "  Optimized FLOP count:  %.3e\n" % opt_cost
+    path_print += "   Theoretical speedup:  %3.3f\n" % (naive_cost / float(opt_cost))
+    path_print += "  Largest intermediate:  %.3e elements\n" % max(size_list) 
+    path_print += "-" * 80 + "\n"
+    path_print += "%6s %6s %24s %40s\n" % header
+    path_print += "-" * 80
+
+    for n, contraction in enumerate(contraction_list):
+        inds, gemm, einsum_str, remaining = contraction
+        remaining_str = ",".join(remaining) + "->" + output_subscript
+        path_run = (scale_list[n], gemm, einsum_str, remaining_str)
+        path_print += "\n%4d    %6s %24s %40s" % path_run
+
+    return (path, path_print)
+
+# Rewrite einsum to handle different cases
+def einsum(*operands, **kwargs):
+    """
+    Evaluates the Einstein summation convention based on the operands,
+    differs from einsum by utilizing intermediate arrays to
+    reduce overall computational time.
+
+    Produces results identical to that of the einsum function; however,
+    the contract function expands on the einsum function by building
+    intermediate arrays to reduce the computational scaling and utilizes
+    BLAS calls when possible.
+
+    Parameters
+    ----------
+    subscripts : str
+        Specifies the subscripts for summation.
+    *operands : list of array_like
+        These are the arrays for the operation.
+    tensordot : bool, optional (default: True)
+        If true use tensordot where possible.
+    path : bool or list, optional (default: ``greedy``)
+        Choose the type of path.
+
+        - if a list is given uses this as the path.
+        - 'greedy' means a N^3 algorithm that greedily
+            chooses the best algorithm.
+        - 'optimal' means a N! algorithm that tries all possible ways of
+            contracting the listed tensors.
+
+    memory_limit : int, optional (default: largest input or output array size)
+        Maximum number of elements allowed in an intermediate array.
+
+
+    Returns
+    -------
+    output : ndarray
+        The results based on Einstein summation convention.
+
+    See Also
+    --------
+    einsum, tensordot, dot
+
+    Notes
+    -----
+    Subscript labels follow the same convention as einsum with the exception
+    that integer indexing and ellipses are not currently supported.
```

Comment:
Is it still the case that ellipses are not supported? I see that they are parsed.